lmdeploy/pytorch_poc/models/functional.py

```diff
@@ -2,17 +2,43 @@
 import math
 from typing import Any, Callable, Optional, Sequence, Tuple
 
+import numpy as np
 import torch
+# import torch.nn.functional as F
 from torch import Tensor
 from torch import distributed as dist
 
 from lmdeploy.pytorch_poc.kernels import (alibi_paged_attention_fwd,
                                           apply_rotary_pos_emb, fill_kv_cache,
-                                          paged_attention_fwd)
+                                          paged_attention_fwd,
+                                          rerope_attention_fwd)
 
 __all__ = ['apply_rotary_pos_emb']
 
 
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., :x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """This is the equivalent of torch.repeat_interleave(x, dim=1,
```

Comment:
So what is the difference between repeat_kv and repeat_interleave?

Reply:
"This is the equivalent of torch.repeat_interleave"